pythonforandroid/bootstrap.py

```diff
@@ -180,14 +197,61 @@ def get_bootstrap_from_recipes(cls, recipes, ctx):
                         ok = False
                         break
                 if ok and bs not in acceptable_bootstraps:
-                    acceptable_bootstraps.append(bs)
+                    acceptable_bootstraps.add(bs)
+
         info('Found {} acceptable bootstraps: {}'.format(
             len(acceptable_bootstraps),
             [bs.name for bs in acceptable_bootstraps]))
-        if acceptable_bootstraps:
-            info('Using the first of these: {}'
-                 .format(acceptable_bootstraps[0].name))
-            return acceptable_bootstraps[0]
+
+        def have_dependency_in_recipes(dep):
+            for dep_list in recipes_with_deps_lists:
+                if dep in dep_list:
+                    return True
+            return False
+
+        # Special rule: return SDL2 bootstrap if there's an sdl2 dep:
+        if (have_dependency_in_recipes("sdl2") and
+                "sdl2" in [b.name for b in acceptable_bootstraps]
+                ):
+            info('Using sdl2 bootstrap since it is in dependencies')
+            return cls.get_bootstrap("sdl2", ctx)
+
+        # Special rule: return "webview" if we depend on common web recipe:
+        for possible_web_dep in known_web_packages:
+            if have_dependency_in_recipes(possible_web_dep):
+                # We have a web package dep!
+                if "webview" in [b.name for b in acceptable_bootstraps]:
+                    info('Using webview bootstrap since common web packages '
+                         'were found {}'.format(
+                             known_web_packages.intersection(recipes)
+                         ))
+                    return cls.get_bootstrap("webview", ctx)
+
+        # Otherwise, go by a default priority ordering to pick best one:
+        def bootstrap_priority(a, b):
+            def rank_bootstrap(bootstrap):
+                """ Returns a ranking index for each bootstrap,
+                    with higher priority ranked with higher number. """
+                if bootstrap.name in default_recipe_priorities:
+                    return default_recipe_priorities.index(bootstrap.name) + 1
+                return 0
+            # Rank bootstraps in order:
+            rank_a = rank_bootstrap(a)
+            rank_b = rank_bootstrap(b)
+            if rank_a != rank_b:
+                return (rank_b - rank_a)
+            else:
+                return (a.name - b.name)  # alphabetic sort for determinism
+
+        prioritized_acceptable_bootstraps = sorted(
+            list(acceptable_bootstraps),
+            key=functools.cmp_to_key(bootstrap_priority)
+        )
```

Comment:
very interesting helper function. I understand you didn't want to expose it because it's a util function only being used in `get_bootstrap_from_recipes()`, but then:
1) it makes the overall `get_bootstrap_from_recipes()` method larger
2) the `bootstrap_priority()` helper is not unit testable

Maybe we should define it as a private top function inside the file `bootstrap.py` file? I have the feeling it would make the overall thing more readable and more testable